examples/showcase-all-resources.tf

```diff
@@ -0,0 +1,105 @@
+# App
+
+resource "ably-app" "app1" {
+  status                 = "enabled"
+  tlsOnly                = "false"
+  fcmKey                 = "AABBQ1KyxCE:APA91bCCYs7r_Q-sqW8HMP_hV4t3vMYx...cJ8344-MhGWODZEuAmg_J4MUJcVQEyDn...I"
+  apnsCertificate        = "-----BEGIN CERTIFICATE-----MIIFaDCC...EXAMPLE...3Dc=-----END CERTIFICATE-----"
+  apnsPrivateKey         = "-----BEGIN PRIVATE KEY-----ABCFaDCC...EXAMPLE...3Dc=-----END PRIVATE KEY-----"
+  apnsUseSandboxEndpoint = false
+}
+
+# Keys
+
+resource "ably-api-key" "api-key-1" {
+  app  = ably-app.app1.app_id
+  name = "KeyName"
+  capability = {
+    channel1 = [""]
+    channel2 = [""]
+  }
+}
+
+# Namespaces
+
+resource "ably-namespace" "namespace1" {
+  app = ably-app.app1.app_id
+  #...
+}
+
+# Rules
+
+resource "ably-rule-source" "example-rule-source-1" {
+  channelFilter = "^my-channel.*"
+  type          = "channel.message"
+}
+
+resource "ably-http-rule-target" "example-http-rule-target" {
+  url = "https://example.com/webhooks"
+  headers = [
+    {
+      name  = "User-Agent",
+      value = "user-agent-string"
+    },
+    {
+      name  = "headerName",
+      value = "headerValue"
+    }
+  ]
+  signingKeyId = "bw66AB"
+  enveloped    = true
+  format       = "json"
+}
```

Comment:
Interested to know why you split out the rule source and rule target. 
Each of these would need a unique identifier to be able to reference them from ably_rule resource. Is that possible and do these expose a unique identifier?

Reply:
This is a good point. The reason I have split them was more UX driven rather than Control API driven. 

Looking for idiomatic examples, I basically tried to follow the model of different AWS related Terraform resources where you have very small atomic resources that link to each other via different forms of references, i.e. 

```
resource "aws_s3_bucket" "bucket-test" {
  bucket = "s3-test-neatly-handy-cow"
}


resource "aws_s3_bucket_versioning" "bucket-test-versioning" {
  bucket = aws_s3_bucket.bucket-test.id
  versioning_configuration {
    status = "Enabled"
  }
}

resource "aws_s3_bucket_public_access_block" "bucket-test-public-access" {
  bucket = aws_s3_bucket.bucket-test.id
  block_public_acls = true
  ignore_public_acls = true
  restrict_public_buckets = true
  block_public_policy = true
}

resource "aws_s3_bucket_logging" "logging-test" {
  bucket = aws_s3_bucket.bucket-test.bucket

  target_bucket = aws_s3_bucket.bucket-test.id
  target_prefix = "log-s3-test-neatly-handy-cow/"
}
```

My understanding is that in this situation our provider implementation would "combine" these separate resources into single rule definitions. The way it would do state management is just by comparing the state of the "combined" rule entity, not by comparing `source` and `target` individually. This should be ok, as in case where either `source` or `target` were changed, we probably want to recreate the entire rule (however, I guess this is also something to decide from the engineering perspective - do we want to treat entities as immutable, as it will make our life easier in many ways, or do we want to actually make updates).

Reply:
Just to be clear though - if the idea that I have just described requires too much extra code I think we can ignore these considerations for now, and present Terraform resources as basically 1 to 1 mapping to our Control API resources. I don't think this will impact UX in any significant way, and we can always improve (and add complexity to implementation) in the future versions.

Reply:
Summary from our call earlier with @graham-russell and @Morganamilo - most likely just having different blocks for different rule types, all within one `ably_rule` resource will just be much easier to understand, more readable, and is still idiomatic as some big providers use this approach (i.e. https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/application_gateway)

@graham-russell will think more about this today and we will make the final decision then.

Reply:
@graham-russell if you had a chance to think about this - do you mind just taking over this branch and adjusting this PR to your thinking?

Reply:
Sure thing @kavalerov

Reply:
So after much deliberation and experimenting, I think the way in which you originally structured is best @kavalerov. 
I tried a single rule resource and personally think that it's too large a potential resource and would add cognitive dissonance during onboarding.

So structure wise for rules, I vote for:
- ably_rule
- ably_rule_source
- ably_rule_target_{target type}. e.g. ably_rule_target_aws_sqs

Note that I've modified the Ably Rule Target resources naming structure to make for cleaner listing in alphabetical order. I think we should follow the convention of `{provider}_{resource_type}_{optional_resource_type_subdivision}`. It is now `ably_rule_target_{target type}` and was previously `ably_{target type}_rule_target`. It'll make listing on the TF registry better. 

Example of the AWS Provider
<img width="395" alt="Screenshot 2022-07-25 at 15 09 03" src="https://user-images.githubusercontent.com/97962881/180797342-8e0d9c5d-cb03-4a57-87df-fd30ed9398ed.png">

Reply:
Sounds good to me - thanks for looking into this!

Reply:
Resolving this now as we have solved this.